File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& mat) {
        int n = mat.size(), m = mat[0].size();
        for(int i=0; i<n; i++) {
            int flag = 0;
            for(int j=0; j<m; j++) {
                int top = i == 0 ? 0 : mat[i-1][j];
                int left = j == 0 ? 0 : mat[i][j-1];
                if(i == 0 && j == 0 && mat[i][j] == 0) mat[i][j] = 1;
                else mat[i][j] = mat[i][j] == 1 ? 0 : top + left;
                
                if(mat[i][j] > 0) flag = 1;
            }
            if(flag == 0) return 0;
        }
        return mat[n-1][m-1];
    }
};

File: 0063-unique-paths-ii/0063-unique-paths-ii.java
class Solution {
    public int uniquePathsWithObstacles(int[][] mat) {
        int n = mat.length, m = mat[0].length;
        int[][] dp = new int[n][m];
        for(int i=0; i<n; i++) {
            if(mat[i][0] == 1) break;
            dp[i][0] = 1;
        }
        for(int j=0; j<m; j++) {
            if(mat[0][j] == 1) break;
            dp[0][j] = 1;
        }
        
        for(int i=1; i<n; i++) {
            for(int j=1; j<m; j++) {
                if(mat[i][j] == 1) dp[i][j] = 0;
                else dp[i][j] = dp[i-1][j] + dp[i][j-1];
            }
        }
        return dp[n-1][m-1];
    }
}